docs/proposals/edgemesh-design.md

```diff
@@ -76,6 +76,113 @@ Add two endpoints definition:
+
+Each provider that wants to register as a Source should implement following interfaces.
+
+```go
+type Source interface {
+	// Name Returns name of the provider
+	Name() string
+	// RegisterListener is used to register listener for a rule with this source type
+	RegisterListener(rule model.Rule, res model.Map, handler func(interface{})) error
+	// UnregisterListener is used to unregister a listener when rule is deleted
+	UnregisterListener(rule model.Rule, res map[string]interface{})
+	// Callback is function for sending response/ACK if required for a request 
+	Callback(map[string]interface{})
+}
+```
+
+```go
+type SourceFactory interface {
+	// Type returns type of the source
+	Type() string
+	// GetSource returns Source implementation
+	GetSource(ep *model.Endpoint) Source
+}
+```
+
+### Target Interfaces
+
+Each provider that wants to register as a Target should implement following interfaces.
+
+```go
+type Target interface {
+	// Name returns name of the provider
+	Name() string
+	// Forward is used to forward the request to target 
+	Forward(map[string]interface{}, interface{}, func(map[string]interface{})) error
+}
+```
+
+```go
+type TargetFactory interface {
+	// Type returns type of the target
+	Type() string
+	// GetTarget returns Target implementation
+	GetTarget(ep *model.Endpoint) Target
+}
+```
+
+### Rule Definition
+
+```go
+// Rule define message translate from Endpoint to Endpoint
+type Rule struct {
+	ID              string    `orm:"column(id); size(36); pk" json:"id" yaml:"id"`
+	Name            string    `orm:"column(name); size(64)" json:"name" yaml:"name"`
+	ProjectID       string    `orm:"column(project_id); size(36); index" json:"project_id" yaml:"project_id"`
+	DomainID        string    `orm:"column(domain_id); size(36); index" json:"domain_id" yaml:"domain_id"`
+	Source          *Endpoint `orm:"column(source); rel(fk)" json:"source" yaml:"source"`
+	SourceResource  Map       `orm:"column(source_resource); size(1024)" json:"source_resource" yaml:"source_resource"`
+	Target          *Endpoint `orm:"column(target); rel(fk)" json:"target" yaml:"target"`
+	TargetResource  Map       `orm:"column(target_resource); size(1024)" json:"target_resource" yaml:"target_resource"`
+	InUsing         bool      `orm:"column(in_using); default(true)" json:"in_using" yaml:"in_using"`
+	IsDeleted       bool      `orm:"column(is_deleted); default(false)" json:"is_deleted" `
+	CreatedAt       time.Time `orm:"column(created_at); auto_now_add; type(datetime)" json:"created_at" `
+	UpdatedAt       time.Time `orm:"column(updated_at); auto_now; type(datetime)" json:"updated_at" `
+	SuccessMessages uint64    `orm:"column(success_messages); default(0)" json:"success_messages" `
+	FailMessages    uint64    `orm:"column(fail_messages); default(0)" json:"fail_messages" `
+	Error           []*Error  `orm:"reverse(many)" json:"-"`
+}
+```
+
+
+### Endpoint Definition
+
+```go
+// Endpoint define how to access external service
+type Endpoint struct {
```

Comment:
can we have comments on top of the member variables of the endpoint to explain what is it about ?

Reply:
Done, Thanks